applications/nrf5340_audio/src/modules/sd_card_playback.c

```diff
@@ -0,0 +1,362 @@
+	if (ret < 0) {
+		LOG_ERR("Error when trying to peek at file segment on SD card. Return value: %d",
+			ret);
+		return ret;
+	}
+
+	return 0;
+}
+
+static int sd_card_playback_thread_play_lc3(void)
+{
+	int ret;
+
+	ret = sd_card_playback_header_read(playback_file_name);
+	if (ret < 0) {
+		LOG_ERR("Audio Lc3 header read failed. Return value: %d", ret);
+		return ret;
+	}
+
+	pcm_frame_size = 2 * lc3_file_header.sample_rate * lc3_file_header.frame_ms / 1000;
+	lc3_playback_cfg.lc3_frames_num =
+		2 * ((lc3_file_header.signal_len_msb << 16) + lc3_file_header.signal_len_lsb) /
+		pcm_frame_size;
+
+	uint8_t pcm_mono_frame[pcm_frame_size];
+
+	sd_card_playback_active = true;
+	for (uint32_t i = 0; i < lc3_playback_cfg.lc3_frames_num; i++) {
+		/* Skip the frame length info to get to the audio data */
+		ret = sd_card_read((char *)&lc3_playback_cfg.lc3_frame_length,
+				   &lc3_playback_cfg.lc3_frame_length_size, &f_seg_read_entry);
+		if (ret < 0) {
+			LOG_ERR("Error when trying to skip file on SD card. Return value: %d", ret);
+			return ret;
+		}
+		uint8_t lc3_frame[lc3_playback_cfg.lc3_frame_length];
+		size_t lc3_fr_len = lc3_playback_cfg.lc3_frame_length;
+		/* Read the audio data frame to be encoded */
+		ret = sd_card_read((char *)lc3_frame, &lc3_fr_len, &f_seg_read_entry);
+		if (ret < 0) {
+			LOG_ERR("Something went wrong when reading from SD card. Return value: %d",
+				ret);
+			return ret;
+		}
+		/* Decode audio data frame */
+		ret = sw_codec_lc3_dec_run((char *)lc3_frame, lc3_playback_cfg.lc3_frame_length,
+					   pcm_frame_size, 1, pcm_mono_frame, &pcm_mono_write_size,
+					   false);
+		if (ret < 0) {
+			LOG_ERR("Error when running decoder. Return value: %d\n", ret);
+			return ret;
+		}
+		/* Wait until there is enough space in the ringbuffer */
+		k_sem_take(&m_sem_load_from_buf_lc3, K_FOREVER);
+		sd_card_playback_buffer_to_ringbuffer((char *)pcm_mono_frame, pcm_mono_write_size);
+	}
+	ret = sd_card_close(&f_seg_read_entry);
+	if (ret < 0) {
+		LOG_ERR("Error when closing file. Return value: %d", ret);
+		return ret;
+	}
+	sd_card_playback_active = false;
+	return 0;
+}
+
+static int sd_card_playback_thread_play_wav(void)
+{
+	int ret;
+
+	size_t pcm_wav_mono_frame_size = wav_playback_cfg.frame_duration_ms *
+					 wav_playback_cfg.sample_rate * wav_playback_cfg.bit_depth /
+					 8;
+	pcm_frame_size = pcm_wav_mono_frame_size;
+	uint8_t pcm_wav_mono_frame[pcm_wav_mono_frame_size];
+	ret = sd_card_open(playback_file_name, &f_seg_read_entry);
+	if (ret < 0) {
+		LOG_ERR("Error when trying to open file on SD card. Return value: %d", ret);
+		return ret;
+	}
+
+	sd_card_playback_active = true;
+	while (sd_card_playback_active) {
+
+		ret = sd_card_read(pcm_wav_mono_frame, &pcm_wav_mono_frame_size, &f_seg_read_entry);
+		if (ret < 0) {
+			LOG_ERR("Error when trying to read file on SD card. Return value: %d", ret);
+			return ret;
+		}
+		if (pcm_wav_mono_frame_size < pcm_frame_size) {
```

Comment:
It checks whether it is able to fill the 1ms pcm buffer with data from the file. If it is not, then it means that we have reached the end of the file and this can be the last iteration